src/CaptureService/include/CaptureService/ClientCaptureEventCollectorBuilderImpl.h

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_SERVICE_CLIENT_CAPTURE_EVENT_COLLECTOR_BUILDER_IMPL_H_
+#define CAPTURE_SERVICE_CLIENT_CAPTURE_EVENT_COLLECTOR_BUILDER_IMPL_H_
+
+#include <grpcpp/grpcpp.h>
+
+#include "CaptureService/ClientCaptureEventCollectorBuilder.h"
+#include "GrpcProtos/services.grpc.pb.h"
+#include "GrpcProtos/services.pb.h"
+#include "ProducerEventProcessor/ClientCaptureEventCollector.h"
+#include "ProducerEventProcessor/GrpcClientCaptureEventCollector.h"
+
+namespace orbit_capture_service {
+
+class ClientCaptureEventCollectorBuilderImpl : public ClientCaptureEventCollectorBuilder {
```

Comment:
Thanks for the suggestion! 

We thought about this before. But in this case, when the cloud collector includes `ClientCaptureEventCollectorBuilder`, it will need the knowledge of `ServerReaderWriter`. We had similar discussion when we decided not to use the previous deprecated plan.

I would like to also hear @beckerhe's idea on this point.

Reply:
So it's a matter of dependencies? But wouldn't the cloud collector end up depending on gRPC anyway if it depends on this module and `ClientCaptureEventCollectorBuilderImpl` has knowledge of gRPC?

Reply:
I like the idea of having factory functions, but what about making them free functions instead?

I thought about something like this structure:
ClientCaptureEventCollectorBuilder.h:
```
class ClientCaptureEventCollectorBuilder { ... };
```

GrpcClientCaptureEventCollectorBuilder.h:
```
std::unique_ptr<ClientCaptureEventCollectorBuilder> CreateGrpcClientCaptureEventCollectorBuilder();
```

GrpcClientCaptureEventCollectorBuilder.cpp:
```
class GrpcClientCaptureEventCollectorBuilder : public ClientCaptureEventCollectorBuilder { ... };

std::unique_ptr<ClientCaptureEventCollectorBuilder> CreateGrpcClientCaptureEventCollectorBuilder() { ... }
```

WhateverClientCaptureEventCollectorBuilder.h:
```
std::unique_ptr<ClientCaptureEventCollectorBuilder> CreateWhateverClientCaptureEventCollectorBuilder();
```

WhateverClientCaptureEventCollectorBuilder.cpp:
```
class WhateverClientCaptureEventCollectorBuilder : public ClientCaptureEventCollectorBuilder { ... };

std::unique_ptr<ClientCaptureEventCollectorBuilder> CreateWhateverClientCaptureEventCollectorBuilder() { ... }
```

Reply:
```
So it's a matter of dependencies? But wouldn't the cloud collector end up depending on gRPC anyway if it depends on this module and ClientCaptureEventCollectorBuilderImpl has knowledge of gRPC?
```

There was an idea to maybe split up this module to reflect the dependencies properly. Not really relevant for the cloud collector because it needs gRPC anyway (for other purposes). More a question of proper software architecture.

Reply:
BTW, since we are now designing a builder for a builder we could also go with a `std::function` that returns a `unique_ptr` to a `ClientCaptureEventCollector`. Simpler in its design, but less consistent because we can't use that for the Waiter below since it has two member functions. Let me know what you think.

Reply:
> There was an idea to maybe split up this module to reflect the dependencies properly.

Eh, I wanted to say that it looks like we would need another module, but I didn't want to go there... 

> we could also go with a std::function that returns a unique_ptr to a ClientCaptureEventCollector

Yep, I had thought about that too, simpler but less consistent. I'm fine with both.

Reply:
Thanks for the suggestion!

I added two free factory functions 
```cpp
std::unique_ptr<ClientCaptureEventCollectorBuilder> CreateGrpcClientCaptureEventCollectorBuilder(
    grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse, orbit_grpc_protos::CaptureRequest>*
        reader_writer);

std::shared_ptr<StartStopCaptureRequestWaiter> CreateGrpcStartStopCaptureRequestWaiter(
    grpc::ServerReaderWriter<orbit_grpc_protos::CaptureResponse, orbit_grpc_protos::CaptureRequest>*
        reader_writer);

```